### PR TITLE
hubble: Rate limit "hubble events queue is full" logs

### LINKED ISSUE #19202: "hubble events queue is full" is flooding logs
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Cilium logs are flooded with "hubble events queue is full" warnings. These should be rate-limited and/or aggregated for longer time period than the until the consumer recovers. Should also consider if this should be at warning level, or if info or debug would be sufficient.

One downside of this that these messages make it harder to see more relevant warnings in the logs, e.g.:
```
Top 5 errors/warnings:
hubble events queue is full; dropping messages
hubble events queue is processing messages again: 1 messages were lost
hubble events queue is processing messages again: 2 messages were lost
hubble events queue is processing messages again: 3 messages were lost
hubble events queue is processing messages again: 4 messages were lost
```

### Cilium Version

master

### Kernel Version

N/A

### Kubernetes Version

N/A

### Sysdump

_No response_

### Relevant log output

```shell
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 12 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 1 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 3 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 7 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 12 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 4 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 1 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 202 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 4 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 1 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 8 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 3 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 1 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 1 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 1 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 1 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 7 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 7 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 1 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 1 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 1 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 1 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 8 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 6 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 1 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 1 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 39 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 8 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 6 messages were lost" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is full; dropping messages" subsys=hubble
Mar 20 19:04:19 runtime cilium-agent[33599]: level=warning msg="hubble events queue is processing messages again: 4 messages were lost" subsys=hubble
...
```

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by ghost**
If Hubble events queue can't keep up with the incoming events, there are two warnings logged:
```
hubble events queue is full; dropping messages
hubble events queue is processing messages again: N messages were lost
```
They are reported to be very noisy in some environments, e.g. with frequent traffic bursts. This PR mitigates it by:
* rate limiting the first message (and making it more helpful)
* reducing the level of the second message from warning to info

This is a rough patch for the problem. The underlying reasons for excessive logging are somewhat different issues:
1. The events queue is considered "back to normal" after only one event is received. That causes the pair of warnings logged again and again in a loop.
1. When Hubble already can't keep up with the incoming messages, it gets additional `LostEvent` messages. That seems... *suboptimal*.
1. Logs and Hubble events are not good signals for reporting the number of lost events. It should be a metric.

A "proper fix" was proposed in https://github.com/cilium/cilium/issues/15112. But rate limiting "hubble events queue is full" warnings should be useful nonetheless. 

Fixes #19202

```release-note
Rate limit "hubble events queue is full" logs
```

### CODE REVIEW

**Review by kaworu — APPROVED** [review]
Thanks for the patch @lambdanis!

Added @joestringer to review because of the discussion at https://github.com/cilium/cilium/issues/3323

File: pkg/hubble/monitor/consumer.go [hunk 1]
```diff
@@ -90,8 +93,8 @@ func (c *consumer) sendEvent(event *observerTypes.MonitorEvent) {
 func (c *consumer) logStartedDropping() {
 	c.lostLock.Lock()
 	defer c.lostLock.Unlock()
-	if c.numEventsLost == 0 {
-		c.observer.GetLogger().Warning("hubble events queue is full; dropping messages")
+	if c.numEventsLost == 0 && c.logLimiter.Allow() {
+		c.observer.GetLogger().Warning("hubble events queue is full: dropping messages; consider increasing the queue size (hubble-event-queue-size) or provisioning more CPU")
```
Comment (rolinh) [inline pkg/hubble/monitor/consumer.go:97]: That's a much more useful log message, thanks!

**Review by rolinh — APPROVED** [review]
:shipit:

**Comment by lambdanis** [comment]
The test failure is this: https://github.com/cilium/cilium/issues/22931

/ci-aks

**Comment by joestringer** [comment]
Thanks, looks like this is passing CI and reviews are in. Should we mark this for backport to v1.13 for the upcoming release? Seems like it could be a nice QoL improvement.

**Comment by epk** [comment]
Is it possible to backport this to v1.11.x?

**Comment by michi-covalent** [comment]
sounds fine to me. @rolinh added `needs-backport` for v1.13 and v1.12, but not for v1.11. let's check with him if it was intentional 👀

**Comment by rolinh** [comment]
> sounds fine to me. @rolinh added `needs-backport` for v1.13 and v1.12, but not for v1.11. let's check with him if it was intentional eyes

That was intentional as when I did add the labels, Cilium v1.12 was the latest stable release and v1.13 had already branched. I was just applying the usual [backporting criteria](https://docs.cilium.io/en/stable/contributing/release/backports/#backport-criteria). I think the patch is low-risk enough that we could backport to v1.11 although that's not something we do usually.

**Comment by isugimpy** [comment]
This doesn't appear to have actually been backported into 1.13.  Is that intentional?  Looking at the releases I only see it landing in 1.12.6, and this just happened to me in production a few minutes ago on 1.13.3.

**Pull request merged**